selene-lib/src/rules/mismatched_arg_count.rs

```diff
@@ -0,0 +1,425 @@
+                )
+            })
+            .collect()
+    }
+
+    fn severity(&self) -> Severity {
+        Severity::Error
+    }
+
+    fn rule_type(&self) -> RuleType {
+        RuleType::Correctness
+    }
+}
+
+struct MismatchedArgCount {
+    parameter_count: ParameterCount,
+    num_provided: PassedArgumentCount,
+    call_range: (usize, usize),
+    function_definition_range: (usize, usize),
+}
+
+#[derive(Clone, Copy, Debug)]
+enum ParameterCount {
+    /// A fixed number of parameters are required: `function(a, b, c)`
+    Fixed(usize),
+    /// Some amount of fixed parameters are required, and the rest are variable: `function(a, b, ...)`
+    Minimum(usize),
+    /// A variable number of parameters can be provided: `function(...)`
+    Variable,
+}
+
+impl ParameterCount {
+    /// Calculates the number of required parameters that must be passed to a function
+    pub fn from_function_body(function_body: &ast::FunctionBody) -> Self {
+        let mut necessary_params = 0;
+
+        for parameter in function_body.parameters() {
+            match parameter {
+                ast::Parameter::Name(_) => necessary_params += 1,
+                ast::Parameter::Ellipse(_) => {
+                    if necessary_params == 0 {
+                        return Self::Variable;
+                    } else {
+                        return Self::Minimum(necessary_params);
+                    }
+                }
+            }
+        }
+
+        Self::Fixed(necessary_params)
+    }
+
+    /// Checks the provided number of arguments to see if it satisfies the number of arguments required
+    /// We will only lint an upper bound. If we have a function(a, b, c) and we call foo(a, b), this will
+    /// pass the lint, since the `nil` could be implicitly provided.
+    pub fn correct_num_args_provided(&self, provided: PassedArgumentCount) -> bool {
+        match self {
+            ParameterCount::Fixed(required) => match provided {
+                PassedArgumentCount::Fixed(provided) => provided <= *required,
+                // If we have function(a, b, c), but we provide foo(a, call()), we cannot infer anything
+                // but if we provide foo(a, b, c, call()), we know we have too many
+                PassedArgumentCount::Variable(atleast_provided) => atleast_provided <= *required,
+            },
+            // function(a, b, ...) - if we call it through foo(a), b and the varargs could be implicitly nil.
+            // there is no upper bound since foo(a, b, c, d) is valid - therefore any amount of arguments provided is valid
+            ParameterCount::Minimum(_) => true,
+            // Any amount of arguments could be provided
+            ParameterCount::Variable => true,
+        }
+    }
+
+    pub fn to_message(&self, provided: PassedArgumentCount) -> String {
+        match self {
+            ParameterCount::Fixed(required) => {
+                format!("{} arguments but {} were supplied", required, provided)
```

Comment:
For these, make sure it reads correctly with 1 argument--i.e. "expected 1 arguments" doesn't read properly.

I suspect we will need this again, so should be fine to just have as a helper function in its own module.